File: crates/rnvim/src/deploy.rs
//! Remote agent deployment over ssh.
//!
//! Preference order:
//! 1. remote platform == local platform → push our own binary
//! 2. prebuilt agent for the remote target, fetched from this version's
//!    GitHub release into ~/.rnvim/dist/ (authenticated locally via `gh`;
//!    the remote never needs GitHub access) → push it
//!
//! Cross-platform deploys therefore require this version's release to be
//! published (release-train discipline); same-platform pushes and `local:`
//! loopback sessions never need one.
//!
//! Artifacts are version-stamped filenames under ~/.rnvim/bin on the remote,
//! so a pure existence check is enough — client and agent can never skew.

use std::io::Write;
use std::path::PathBuf;
use std::process::{Command, Stdio};

use anyhow::{bail, Context, Result};

const REPO: &str = "beardnick/rnvim";

struct Probe {
    uname: String, // e.g. "Linux x86_64" / "Darwin arm64"
    have_bin: bool,
}

fn bin_path() -> String {
    format!("$HOME/.rnvim/bin/rnvim-agent-{}", env!("CARGO_PKG_VERSION"))
}

fn ssh_run(host: &str, script: &str, stdin_data: Option<&[u8]>) -> Result<String> {
    let mut cmd = Command::new("ssh");
    cmd.args(["-o", "BatchMode=yes", host, script])
        .stdin(if stdin_data.is_some() {
            Stdio::piped()
        } else {
            Stdio::null()
        })
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());
    let mut child = cmd.spawn().context("spawn ssh")?;
    if let Some(data) = stdin_data {
        child
            .stdin
            .take()
            .context("ssh stdin")?
            .write_all(data)
            .context("stream to ssh")?;
        // stdin drops here, closing the pipe so the remote `cat` finishes
    }
    let out = child.wait_with_output().context("wait ssh")?;
    if !out.status.success() {
        bail!(
            "ssh to {host} failed: {}\n{}",
            out.status,
            String::from_utf8_lossy(&out.stderr).trim()
        );
    }
    Ok(String::from_utf8_lossy(&out.stdout).into_owned())
}

fn probe(host: &str) -> Result<Probe> {
    let script = format!(
        "uname -sm; test -x {bin} && echo bin=yes || echo bin=no",
        bin = bin_path(),
    );
    let out = ssh_run(host, &script, None)?;
    let mut lines = out.lines();
    let uname = lines.next().unwrap_or_default().trim().to_string();
    if uname.is_empty() {
        bail!("could not probe {host}: empty uname output");
    }
    let rest: Vec<&str> = lines.map(str::trim).collect();
    Ok(Probe {
        uname,
        have_bin: rest.contains(&"bin=yes"),
    })
}

/// Local platform in `uname -sm` terms, for comparison with the remote.
fn local_uname_sm() -> String {
    let os = match std::env::consts::OS {
        "macos" => "Darwin",
        "linux" => "Linux",
        other => other,
    };
    let arch = match (std::env::consts::OS, std::env::consts::ARCH) {
        ("macos", "aarch64") => "arm64",
        (_, "aarch64") => "aarch64",
        (_, a) => a,
    };
    format!("{os} {arch}")
}

/// Rust target triple for a remote's `uname -sm`, matching release assets.
fn remote_target(uname_sm: &str) -> Option<&'static str> {
    match uname_sm {
        "Linux x86_64" => Some("x86_64-unknown-linux-musl"),
        "Linux aarch64" => Some("aarch64-unknown-linux-musl"),
        "Darwin arm64" => Some("aarch64-apple-darwin"),
        "Darwin x86_64" => Some("x86_64-apple-darwin"),
        _ => None,
    }
}

/// Fetch the prebuilt agent for `target` from this version's GitHub release,
/// caching it under ~/.rnvim/dist/<version>/. The cache also makes previously
/// used targets available offline.
fn fetch_agent_dist(target: &str) -> Result<PathBuf> {
    let version = env!("CARGO_PKG_VERSION");
    let dist_dir = crate::nvim::rnvim_home()?.join("dist").join(version);
    let asset = format!("rnvim-{target}");
    let path = dist_dir.join(&asset);
    if path.exists() {
        return Ok(path);
    }
    std::fs::create_dir_all(&dist_dir)?;
    let tag = format!("v{version}");

    // `gh` carries auth for the private repo; anonymous curl works once the
    // repo is public. Either way the download happens locally, never remotely.
    let gh_ok = Command::new("gh")
        .args([
            "release",
            "download",
            &tag,
            "--repo",
            REPO,
            "--pattern",
            &asset,
            "--dir",
        ])
        .arg(&dist_dir)
        .status()
        .map(|s| s.success())
        .unwrap_or(false);
    if !gh_ok {
        let url = format!("https://github.com/{REPO}/releases/download/{tag}/{asset}");
        let curl_ok = Command::new("curl")
            .args(["-fsSL", "-o"])
            .arg(&path)
            .arg(&url)
            .status()
            .map(|s| s.success())
            .unwrap_or(false);
        if !curl_ok {
            let _ = std::fs::remove_file(&path);
            bail!("no prebuilt agent {asset} for release {tag} (gh and curl both failed)");
        }
    }
    if !path.exists() {
        bail!(
            "download reported success but {} is missing",
            path.display()
        );
    }
    Ok(path)
}

fn push_agent_binary(host: &str, data: &[u8]) -> Result<()> {
    let script = format!(
        "mkdir -p $HOME/.rnvim/bin && cat > {bin}.tmp && chmod +x {bin}.tmp && mv {bin}.tmp {bin}",
        bin = bin_path()
    );
    ssh_run(host, &script, Some(data))?;
    Ok(())
}

/// Make sure a compatible agent exists on the remote; return the shell
/// command that starts it (passed to ssh by the transport).
pub fn ensure_remote_agent(host: &str) -> Result<String> {
    let p = probe(host)?;
    let bin_cmd = format!("{} agent --stdio", bin_path());

    if p.have_bin {
        return Ok(bin_cmd);
    }

    // 1. Same platform: our own binary IS the agent.
    if p.uname == local_uname_sm() {
        eprintln!("[rnvim] deploying agent binary to {host} ({})...", p.uname);
        let exe = std::env::current_exe().context("current_exe")?;
        let data = std::fs::read(&exe).context("read own binary")?;
        push_agent_binary(host, &data)?;
        return Ok(bin_cmd);
    }

    // 2. Prebuilt agent from this version's release.
    let Some(target) = remote_target(&p.uname) else {
        bail!(
            "remote {host} is {} — no agent build exists for this platform \
             (supported: Linux/macOS on x86_64/aarch64)",
            p.uname
        );
    };
    let dist = fetch_agent_dist(target).with_context(|| {
        format!(
            "no agent available for {host} ({}): the v{} release must be published \
             for cross-platform deploys",
            p.uname,
            env!("CARGO_PKG_VERSION")
        )
    })?;
    eprintln!("[rnvim] deploying prebuilt agent ({target}) to {host}...");
    let data = std::fs::read(&dist).with_context(|| format!("read {}", dist.display()))?;
    push_agent_binary(host, &data)?;
    Ok(bin_cmd)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn remote_targets_map_to_release_assets() {
        assert_eq!(
            remote_target("Linux x86_64"),
            Some("x86_64-unknown-linux-musl")
        );
        assert_eq!(
            remote_target("Linux aarch64"),
            Some("aarch64-unknown-linux-musl")
        );
        assert_eq!(remote_target("Darwin arm64"), Some("aarch64-apple-darwin"));
        assert_eq!(remote_target("Darwin x86_64"), Some("x86_64-apple-darwin"));
        assert_eq!(remote_target("SunOS sparc64"), None);
    }
}

File: lua/rnvim/deploy.lua
-- Remote agent deployment over ssh (a Lua port of the old deploy.rs).
--
-- Preference order:
-- 1. version-stamped agent already on the remote → run it
-- 2. prebuilt agent for the remote target, fetched from this version's
--    GitHub release into ~/.rnvim/dist/ (authenticated locally via `gh`
--    when available; anonymous curl once the repo is public) → push it
--
-- Artifacts are version-stamped filenames under ~/.rnvim/bin on the
-- remote, so a pure existence check is enough — plugin and agent can
-- never skew.

local util = require("rnvim.util")

local M = {}

local REPO = "beardnick/rnvim"

local function bin_path()
  return "$HOME/.rnvim/bin/rnvim-agent-" .. util.agent_version()
end

local function ssh_run(host, script, stdin_data)
  local res = vim
    .system({ "ssh", "-o", "BatchMode=yes", host, script }, { stdin = stdin_data, text = true })
    :wait()
  if res.code ~= 0 then
    error(("[rnvim] ssh to %s failed: %s"):format(host, (res.stderr or ""):gsub("%s+$", "")))
  end
  return res.stdout or ""
end

--- Download the prebuilt agent for `target`, caching it under
--- ~/.rnvim/dist/<version>/. Returns the local path.
local function fetch_agent_dist(target)
  local version = util.agent_version()
  local dist_dir = util.home("dist", version)
  local asset = "rnvim-agent-" .. target
  local path = vim.fs.joinpath(dist_dir, asset)
  if vim.uv.fs_stat(path) then
    return path
  end
  util.ensure_dir(dist_dir)
  local tag = "v" .. version

  if vim.fn.executable("gh") == 1 then
    local res = vim
      .system({ "gh", "release", "download", tag, "--repo", REPO, "--pattern", asset, "--dir", dist_dir })
      :wait()
    if res.code == 0 and vim.uv.fs_stat(path) then
      return path
    end
  end
  local url = ("https://github.com/%s/releases/download/%s/%s"):format(REPO, tag, asset)
  local res = vim.system({ "curl", "-fsSL", "-o", path .. ".part", url }):wait()
  if res.code ~= 0 or not vim.uv.fs_stat(path .. ".part") then
    error(
      ("[rnvim] no prebuilt agent %s for release %s — the release must be published (gh and curl both failed)"):format(
        asset,
        tag
      )
    )
  end
  assert(vim.uv.fs_rename(path .. ".part", path))
  return path
end

--- Agent binary for `local:` loopback sessions: an explicit override
--- (RNVIM_AGENT_BIN / vim.g.rnvim_agent_bin), else the dist cache.
function M.local_agent_bin()
  local target = util.rust_target(util.local_uname_sm())
  if not target then
    error("[rnvim] unsupported local platform: " .. util.local_uname_sm())
  end
  local path = fetch_agent_dist(target)
  vim.uv.fs_chmod(path, 493) -- 0755
  return path
end

local function push_agent_binary(host, local_file)
  local f = assert(io.open(local_file, "rb"))
  local data = f:read("*a")
  f:close()
  local bin = bin_path()
  ssh_run(
    host,
    ("mkdir -p $HOME/.rnvim/bin && cat > %s.tmp && chmod +x %s.tmp && mv %s.tmp %s"):format(bin, bin, bin, bin),
    data
  )
end

--- Make sure a compatible agent exists on `host`; return the shell command
--- that starts it (passed to ssh by the transport).
function M.ensure_remote_agent(host)
  local probe = ssh_run(host, ("uname -sm; test -x %s && echo bin=yes || echo bin=no"):format(bin_path()))
  local lines = vim.split(probe, "\n", { trimempty = true })
  local uname = vim.trim(lines[1] or "")
  if uname == "" then
    error("[rnvim] could not probe " .. host .. ": empty uname output")
  end
  local bin_cmd = bin_path() .. " --stdio"
  if vim.list_contains(lines, "bin=yes") then
    return bin_cmd
  end

  local target = util.rust_target(uname)
  if not target then
    error(
      ("[rnvim] remote %s is %s — no agent build exists for this platform (supported: Linux/macOS on x86_64/aarch64)"):format(
        host,
        uname
      )
    )
  end
  util.notify(("deploying agent (%s) to %s..."):format(target, host))
  push_agent_binary(host, fetch_agent_dist(target))
  return bin_cmd
end

return M
